lib/PreviewProvider.js

```diff
@@ -0,0 +1,137 @@
+/* globals Task */
+"use strict";
+
+const {Cu} = require("chrome");
+const ss = require("sdk/simple-storage");
+const {setTimeout} = require("sdk/timers");
+const {siteFilter} = require("content-src/lib/filters");
+
+const EMBEDLY_PROXY_MAX_LINKS = 25; // number of links embedly proxy accepts per request
+const EMBEDLY_ENDPOINT = "https://embedly-proxy.dev.mozaws.net/v2/extract";
+const CLEAR_CACHE_TIMEOUT = 86400000; // 24 hours to clear cache
+
+Cu.importGlobalProperties(["fetch"]);
+Cu.import("resource://gre/modules/Task.jsm");
+
+function PreviewProvider() {
+  this.init();
+}
+
+PreviewProvider.prototype = {
+  // TODO: santize, filter, and dedupe urls
+  /**
+    * Clean up cache
+    */
+  cleanUpCache: Task.async(function*() {
+    setTimeout(() => {
+      let currentTime = Date.now();
+      Object.keys(ss.storage.embedlyData).forEach(item => {
+        if (currentTime - ss.storage.embedlyData[item].accessTime > CLEAR_CACHE_TIMEOUT) {
+          delete ss.storage.embedlyData[item];
+        }
+      });
+      this.cleanUpCache();
+    }, CLEAR_CACHE_TIMEOUT);
+  }),
+
+  /**
+    * Dedupe urls
+    */
+  dedupeURLs(items) {
+    let createKey =  (item => item);
+    let compare = (() => false);
+    const valueMap = new Map();
+    items.forEach(value => {
+      const key = createKey(value);
+      if (!valueMap.has(key) || compare(valueMap.get(key), value)) {
+        valueMap.set(key, value);
+      }
+    });
+    return Array.from(valueMap.values());
+  },
+
+  /**
+    * Filter cached links out, and store their embedly data
+    */
+  getCachedLinks(links) {
+    let cachedLinks = [];
+    links.forEach(link => {
+      if (link && ss.storage.embedlyData[link.url] && link.url === ss.storage.embedlyData[link.url].url) {
+        cachedLinks.push(ss.storage.embedlyData[link.url]);
+      }
+    });
+    return cachedLinks;
+  },
+
+  /**
+    * Prime the cache with embedly data
+    */
+  primeCache: Task.async(function*(links) {
+    this.saveNewLinks(links);
+  }),
+
+  /**
+   * Filter out new links and request them from embedly
+   */
+  saveNewLinks: Task.async(function*(links) {
+    let newLinks = links.filter(link => link && !ss.storage.embedlyData[link.url]);
+    let requestQueue = [];
+    while (newLinks.length !== 0) {
+      // we have some new links we need to fetch the embedly data for, put them on the queue
+      requestQueue.push(newLinks.splice(0, EMBEDLY_PROXY_MAX_LINKS));
+    }
+    // for each bundle of 25 links, spawn off a new request to embedly
+    requestQueue.forEach(requestBundle => {
+      Task.spawn(function*() {
```

Comment:
Yeah I was thinking of that too. I was concerned on blocking on all the promises to resolve before adding it to the cache. Say if we request 100 links, the first 25 (chunk 1) would resolve first then need to wait until the next 75 (chunks 2,3,4) resolve before being added to the cache, correct?